src/objects.c

```diff
@@ -1393,6 +1393,11 @@ void disconnect_server(PgSocket *server, bool send_term, const char *reason, ...
  */
 void disconnect_client(PgSocket *client, bool notify, const char *reason, ...)
 {
+	if (client->login_user_credentials) {
+		if (client->login_user_credentials->global_user) {
+			client->login_user_credentials->global_user->client_connection_count--;
```

Comment:
This substraction should only be done when this specific clien actually increased the connection count before. 

Right now if a user is at the limit. The next failing connection attempt (due to hitting the limit) will then also decrease the connection count. So the connection attempt after that will succeed again, eventhough we're still at the client connection limit.

Reply:
Okay so I added a unit test assert involving the situation you described above. Moved the increment to right before the initial check and changed the inequality accordingly. Seems to satisfy all test cases though there might be another corner case I'm not seeing.

An alternate way of going about this is adding a boolean to each client that specifies if they contribute to the user level count and check that before decrementing the count. I feel like the logic in this commit is better since it adds less variables, as long as the logic is correct.

Another thing that I notice is that this new logic does not apply to PAM authed users though I feel like it should.

Reply:
> An alternate way of going about this is adding a boolean to each client that specifies if they contribute to the user level count and check that before decrementing the count. I feel like the logic in this commit is better since it adds less variables, as long as the logic is correct.

Agreed.

> Another thing that I notice is that this new logic does not apply to PAM authed users though I feel like it should.

Agreed. Everywhere we set `client->login_user_credentials` we should do this incrementing/decrementing correctly. Also the block below where auth_user is used should do this. I think the easiest way to do this is do

```
client->login_user_credentials = /* something */
if (!check_user_connection_count())
    return;

bool check_user_connection_count() {
    /* your check */
}
```

Reply:
Encapsulated the logic into the function and added the call to everywhere client->login_user_credentials is assigned. Tests all passed but the surface area of the change is now much bigger which may mean more testing is required (limits correctly incremented, decremented and enforced for pam users, admin users, etc). Also the check_user_connection_count function could probably be refactored a bit (lot of nesting).

Reply:
@JelteF Expanded test coverage to include admin users. Refactored the check_user_connection_count function to prefer early returrns instead of nested if statements. Refactored the tests so they should be faster, rely less on asynchronously running pg_sleep.

Could also expand test coverage to stats users as I believe are instantiated differently than admin users. Also missing is test coverage around PAM users though I dont believe there is PAM coverage elsewhere in the test codebase either.

Let me know if anything else needs to be done with this PR in order to get it in a mergable state.

Reply:
If tests pass, I think this is probably good. I'm unavailable for the next ~3 weeks though.

Reply:
Understood, thanks for the heads up.